src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java

```diff
@@ -0,0 +1,457 @@
+        delta.applyTo(result);
+        return null;
+      } catch (PatchFailedException e) {
+        return new PatchFailedException(e.getMessage() +
+            "\nFailed to apply delta:\n    " + delta.toString());
+      }
+    }
+  }
+
+  private enum LineType {
+    OLD_FILE,
+    NEW_FILE,
+    CHUNK_HEAD,
+    CHUNK_ADD,
+    CHUNK_DEL,
+    CHUNK_EQL,
+    GIT_HEADER,
+    RENAME_FROM,
+    RENAME_TO,
+    GIT_LINE,
+    UNKNOWN
+  }
+
+  private static final String[] GIT_LINE_PREFIXES = {
+      "old mode ",
+      "new mode ",
+      "deleted file mode ",
+      "new file mode ",
+      "copy from ",
+      "copy to ",
+      "rename old ",
+      "rename new ",
+      "similarity index ",
+      "dissimilarity index ",
+      "index "
+  };
+
+  private static LineType getLineType(String line, boolean isReadingChunk) {
+    if (isReadingChunk) {
+      if (line.startsWith("+")) {
+        return LineType.CHUNK_ADD;
+      }
+      if (line.startsWith("-")) {
+        return LineType.CHUNK_DEL;
+      }
+      if (line.startsWith(" ") || line.isEmpty()) {
+        return LineType.CHUNK_EQL;
+      }
+    } else {
+      if (line.startsWith("--- ")) {
+        return LineType.OLD_FILE;
+      }
+      if (line.startsWith("+++ ")) {
+        return LineType.NEW_FILE;
+      }
+      if (line.startsWith("diff --git ")) {
+        return LineType.GIT_HEADER;
+      }
+      if (line.startsWith("rename from ")) {
+        return LineType.RENAME_FROM;
+      }
+      if (line.startsWith("rename to ")) {
+        return LineType.RENAME_TO;
+      }
+      for (String prefix : GIT_LINE_PREFIXES) {
+        if (line.startsWith(prefix)) {
+          return LineType.GIT_LINE;
+        }
+      }
+    }
+    if (line.startsWith("@@") && line.lastIndexOf("@@") != 0) {
+      int pos = line.indexOf("@@", 2);
+      Matcher m = chunkHeaderRe.matcher(line.substring(0, pos + 2));
+      if (m.find()) {
+        return LineType.CHUNK_HEAD;
+      }
+    }
+    return LineType.UNKNOWN;
+  }
+
+  @VisibleForTesting
+  public static List<String> readFile(Path file) throws IOException {
+    return Lists.newArrayList(FileSystemUtils.readLines(file, StandardCharsets.UTF_8));
+  }
+
+  private static void writeFile(Path file, List<String> content) throws IOException {
+    FileSystemUtils.writeLinesAs(file, StandardCharsets.UTF_8, content);
+  }
+
+  private static void applyPatchToFile(List<String> patchContent, Path oldFile, Path newFile,
+      boolean isRenaming)
+      throws IOException, PatchFailedException {
+    // Neither of newFile and oldFile can be none if we should do a rename.
+    if (isRenaming && (oldFile == null || newFile == null)) {
+      String msg = "old file is " + (oldFile == null ? "not specified" : oldFile.getPathString());
+      msg += " and new file is " + (newFile == null ? "not specified" : newFile.getPathString());
+      throw new PatchFailedException("Cannot rename file without two valid file names, " + msg);
```

Comment:
This will use the full absolute file path... I would change it to "Can not rename file from XXX: new file name is not specified", "Can not rename file to XXX: existing file name is not specified", with XXX being the relative file path, as it is given in the patch.
Sorry I know it is tiny details...